File: lib/blocks/debounce.c
#include <time.h>
#include "debounce.h"

void block_Debounce(struct Debounce* self)
{
    bool s  = self->input;
    bool r  = self->reset;
    int  d  = self->duration;
    bool c  = self->current;
    int  t  = self->last_state_change;

    bool reset = block_reset();
    int now = clock();
    int duration_elapsed = (now - t) > d;
    bool signal_steady = s == c;

    if (reset)
    {
        struct Debounce init = {};
        *self = init;
    }
    else if (signal_steady && duration_elapsed)
    {
        self->current = s;
        self->output = s;
    }
    else if (signal_steady)
    {
        /* do nothing */
    }
    else
    {
        self->last_state_change = now;
    }
}

File: modules/debounce.m
// -----------------------------------------------------------------------------
// Copyright 25 June 2023
// Author: Preston Langford
// This module debounces a boolean signal
// -----------------------------------------------------------------------------

#include "time.h"
#include "debounce.h"

module Debounce
{
    in bool reset;
    in bool input;
    in i64 duration;

    state i64 now = time();
    state bool unstable = or(reset,xor(input,state));
    state i64 before = mux(unstable,now,before);

    state i64 elapsed = sub(now,before);
    state bool expired = gt(elapsed,duration);

    state bool state = and(
        not(reset),
        mux(expired,input,state)
    );

    out bool output = state;
};